### PR TITLE
Unbox client handshake future

### PR DESCRIPTION
**Pull request by hawkw**
I've rewritten the client handshake to store the inner future with a named type, rather than as a `Box<Future>`. This also means that the `'static` lifetime bounds on the type parameters `T` and `B` can be removed.

I've ran the tests against this branch and everything passes. Since no new functionality was added, I haven't added any additional tests.

Potential drawbacks of this change are hat it does make some code in `client.rs` a bit more complex, and that the type of the inner future in `Handshake` may need to be updated if certain aspects of the implementation of `Client::handshake2` are changed. However, there isn't currently any other way to remove the `Box` in this situation.

### CODE REVIEW

**Comment by hawkw** [comment]
N.B. that I'd also like to unbox the future in `server::Handshake` as well. The way the `Server::handshake2()` is currently implemented (with a `move` closure that captures the value of `framed_write`) complicates this somewhat.

File: src/client.rs [hunk 1]
```diff
@@ -52,26 +58,40 @@ impl<T, B> Client<T, B>
...
-
-                // Send initial settings frame
-                match framed_write.start_send(settings.into()) {
-                    Ok(AsyncSink::Ready) => {
-                        let connection = proto::from_framed_write(framed_write);
-                        Ok(Client { connection })
-                    }
-                    Ok(_) => unreachable!(),
-                    Err(e) => Err(ConnectionError::from(e)),
+        // this is a function so that we can pass a function pointer (rather
```
Comment (seanmonstar) [inline src/client.rs]: FWIW, [as of Rust 1.19](https://blog.rust-lang.org/2017/07/20/Rust-1.19.html), a closure that doesn't capture over its environment can be cast to a function pointer.

Comment (hawkw) [reply src/client.rs]: Ooh, fancy – I must have missed that change.

File: src/client.rs [hunk 2]
```diff
@@ -5,20 +5,27 @@ use proto::{self, Connection, WindowSize};
...
+            MapErr<WriteAll<T, &'static [u8]>, fn(IoError) -> ConnectionError>,
+            Result<Client<T, B>, ConnectionError>,
+            fn((T, &'static [u8])) -> Result<Client<T, B>, ConnectionError>
+        >
+
 }
 
 /// Marker type indicating a client peer
-pub struct Client<T, B: IntoBuf> {
+pub struct Client<T: AsyncRead + AsyncWrite, B: IntoBuf> {
```
Comment (carllerche) [inline src/client.rs]: Adding this bound is fine w/ me. I'm just curious if this became required due to the change or it was added for some other reason.

**Comment by carllerche** [comment]
I'll merge this after #35 lands. It will be easier to fix any merge conflicts here than the other way around...

File: src/client.rs (hunk 2 above)
Comment (hawkw) [reply src/client.rs]: If I recall correctly, this bound was necessary for this change in 658c80b6a9535b4a5645b4c29f5275ca516cec28. I'm not sure if it's still necessary as of a67fd958fdc1977c1d4f8438dd2e898cf78680c5?

**Pull request merged**